src/backend/tcop/utility.c

```diff
@@ -601,91 +602,99 @@ standard_ProcessUtility(PlannedStmt *pstmt,
 			 */
 		case T_TransactionStmt:
 			{
-				TransactionStmt *stmt = (TransactionStmt *) parsetree;
-
-				switch (stmt->kind)
+				if(NestedTranCount > 0 || (sql_dialect == SQL_DIALECT_TSQL && !IsTransactionBlockActive()))
 				{
-						/*
-						 * START TRANSACTION, as defined by SQL99: Identical
-						 * to BEGIN.  Same code for both.
-						 */
-					case TRANS_STMT_BEGIN:
-					case TRANS_STMT_START:
-						{
-							ListCell   *lc;
+					if (transactionStmt_hook)
+						(*transactionStmt_hook)(pstmt, params, qc); 
+				}
+				else
+				{
+					TransactionStmt *stmt = (TransactionStmt *) parsetree;
 
-							BeginTransactionBlock();
-							foreach(lc, stmt->options)
+					switch (stmt->kind)
+					{
+							/*
+							* START TRANSACTION, as defined by SQL99: Identical
+							* to BEGIN.  Same code for both.
+							*/
+						case TRANS_STMT_BEGIN:
+						case TRANS_STMT_START:
 							{
-								DefElem    *item = (DefElem *) lfirst(lc);
-
-								if (strcmp(item->defname, "transaction_isolation") == 0)
-									SetPGVariable("transaction_isolation",
-												  list_make1(item->arg),
-												  true);
-								else if (strcmp(item->defname, "transaction_read_only") == 0)
-									SetPGVariable("transaction_read_only",
-												  list_make1(item->arg),
-												  true);
-								else if (strcmp(item->defname, "transaction_deferrable") == 0)
-									SetPGVariable("transaction_deferrable",
-												  list_make1(item->arg),
-												  true);
+								ListCell   *lc;
+
+								BeginTransactionBlock();
+								foreach(lc, stmt->options)
+								{
+									DefElem    *item = (DefElem *) lfirst(lc);
+
+									if (strcmp(item->defname, "transaction_isolation") == 0)
+										SetPGVariable("transaction_isolation",
+													list_make1(item->arg),
+													true);
+									else if (strcmp(item->defname, "transaction_read_only") == 0)
+										SetPGVariable("transaction_read_only",
+													list_make1(item->arg),
+													true);
+									else if (strcmp(item->defname, "transaction_deferrable") == 0)
+										SetPGVariable("transaction_deferrable",
+													list_make1(item->arg),
+													true);
+								}
 							}
-						}
-						break;
+							break;
 
-					case TRANS_STMT_COMMIT:
-						if (!EndTransactionBlock(stmt->chain))
-						{
-							/* report unsuccessful commit in qc */
-							if (qc)
-								SetQueryCompletion(qc, CMDTAG_ROLLBACK, 0);
-						}
-						break;
+						case TRANS_STMT_COMMIT:
+							if (!EndTransactionBlock(stmt->chain))
+							{
+								/* report unsuccessful commit in qc */
+								if (qc)
+									SetQueryCompletion(qc, CMDTAG_ROLLBACK, 0);
+							}
+							break;
 
-					case TRANS_STMT_PREPARE:
-						if (!PrepareTransactionBlock(stmt->gid))
-						{
-							/* report unsuccessful commit in qc */
-							if (qc)
-								SetQueryCompletion(qc, CMDTAG_ROLLBACK, 0);
-						}
-						break;
+						case TRANS_STMT_PREPARE:
```

Comment:
please make sure we only have diff for actual changes, indentation is creating unnecessary diff